api/nodes/serializers.py

```diff
@@ -124,9 +128,124 @@ def update(self, instance, validated_data):
+        node = self.context['view'].get_node()
+        contributor = get_object_or_error(User, validated_data['_id'])
+        # Node object checks for contributor existence but can still change permissions anyway
+        if contributor in node.contributors:
+            raise exceptions.ValidationError('{} is already a validated contributor'.format(contributor.username))
+
+        bibliographic = validated_data['bibliographic']
+        permissions = self.get_permissions_list(validated_data['permission']) \
+            if 'permission' in validated_data else [osf_permissions.READ, osf_permissions.WRITE]
+        node.add_contributor(contributor=contributor, auth=auth, visible=bibliographic, permissions=permissions, save=True)
+        contributor.permission = node.get_permissions(contributor)[-1]
+        contributor.bibliographic = node.get_visible(contributor)
+        contributor.node_id = node._id
+        return contributor
+
+    @staticmethod
+    def get_permissions_list(permission):
+        if permission == osf_permissions.ADMIN:
+            return [osf_permissions.READ, osf_permissions.WRITE, osf_permissions.ADMIN]
+        elif permission == osf_permissions.WRITE:
+            return [osf_permissions.READ, osf_permissions.WRITE]
+        elif permission == osf_permissions.READ:
+            return [osf_permissions.READ]
+        else:
+            return osf_permissions.DEFAULT_CONTRIBUTOR_PERMISSIONS
+
+
+class NodeContributorDetailSerializer(NodeContributorsSerializer):
+    """ Overrides node contributor serializer to make id read only and add additional methods
+    """
+    id = ser.CharField(read_only=True, source='_id', label='ID')
+
+    def update(self, instance, validated_data):
+        contributor = instance
+        auth = Auth(self.context['request'].user)
+        node = self.context['view'].get_node()
+        if 'permission' in validated_data and validated_data['permission'] != '':
+            self.set_contributor_permissions(node, validated_data['permission'], contributor, auth)
+        try:
+            node.set_visible(contributor, validated_data['bibliographic'], save=True)
+        except ValueError as e:
+            raise exceptions.ValidationError(e)
+        contributor.permission = node.get_permissions(contributor)[-1]
+        contributor.bibliographic = node.get_visible(contributor)
+        contributor.node_id = node._id
+        return contributor
+
+    def set_contributor_permissions(self, node, permission, contributor, auth):
+        """ Set contributor permissions. Checks to make sure unique admin is
+        removing own admin privilege. Based on the ContributorDetailPermissions
+        permissions class, the current user must be an admin.
+        """
+        assert node.has_permission(auth.user, osf_permissions.ADMIN), "Only admins can modify contributor permissions"
+        permissions = self.get_permissions_list(permission)
+        if not self.has_multiple_admins(node):
```

Comment:
I'm thinking this logic probably belongs in `Node#set_permissions`, similar to `Node#set_visible`.